mantidproject/mantid
# Load Missing Default Values from IDF in Reflectometry Settings Tab

This PR adds the ability to load default values for all* properties on the Reflectometry GUI settings tab from the instrument parameter file.

\* Excludes the transmission runs property.

**To test:**
Using the table below as a reference for optional vs mandatory values and their expected default values:
1. For various instruments, including at least one which uses polarisation corrections e.g. `OFFSPEC`
    try setting default values in the IPF and then loading them ensuring that the correct value is set in the settings tab (see below for detailed instructions).
2. For various instruments try removing mandatory parameters from the IPF and then attempting to load the defaults. Ensure that an appropriate error message is generated, listing all of the missing parameters.
3. Ensure that it is possible to load the parameter `I0MonitorIndex` as either an int or a double.
4. Code review

**Setting an IPF parameter value.**
1. Open an instrument's IPF file found at `instrument/INSTRUMENT_NAME_Parameters.xml`
2. Include the snippet below substituting {{...}} for the corresponding name, type or value.
```xml
<parameter name="{{ParameterName}}" type="{{ParameterType}}">
    <value val="{{ParameterValue}}"/>
</parameter>
```
- The type attribute can be omitted and will then be deduced/default to double.

**Loading default values from the Reflectometry Settings Tab**
1. Open the Reflectometry GUI and switch to the settings tab. Pressing `Get Defaults` in either experiment settings or instrument settings loads the relevant parameters from the IPF.
2. The boxes/settings should now be populated and any errors/warnings reported.
#### Instrument Parameters ####

| Parameter                       | Type                | Optional?                  | Default Value  |
| --------------------------------| ------------------ | ---------------------------| ------------------|
| `IntegratedMonitors`        | bool                 | Yes                          | Algorithm        |
| `MonitorIntegralMax`       | double             | No                            | N/A                |
| `MonitorIntegralMin`        | double             | No                            | N/A                |
| `MonitorBackgroundMin`  | double             | No                            | N/A                |
| `MonitorBackgroundMax` | double             | No                            | N/A                |
| `LambdaMin`                   | double             | No                            | N/A                |
| `LambdaMax`                  | double             | No                            | N/A                |
| `I0MonitorIndex`              | double or int    | No                            | N/A                |
| `ProcessingInstructions`   | string              | Yes                           | ""                   |
| `CorrectDetectors`           | bool                | Yes                           | Algorithm       |
| `DetectorCorrectionType`  | string              | Yes                           | Algorithm       |

#### Experiment Parameters ####
| Parameter                       | Type                | Optional?                  | Default Value  |
| --------------------------------| ------------------ | ---------------------------| ------------------|
| `AnalysisMode        `        | string              | Yes                          | Algorithm        |
| `SummationType`             | string              | Yes                          | Algorithm        |
| `ReductionType       `        | string              | Yes                          | Algorithm        |
| `CRho`                            | string              | Yes                          | 1                    |
| `CAlpha`                            | string              | Yes                          | 1                    |
| `CAp`                            | string              | Yes                          | 1                    |
| `CPp`                            | string              | Yes                          | 1                    |
| `dQ/Q`                          | double             | Yes                           | "" |
| `Scale`                          | double             | Yes                          | "" |
| `Stitch1DMany`              | string            | Yes                          | "" |
| `TransRunStartOverlap`  | double           | If on SURF or CRISP | "" if applicable |
| `TransRunEndOverlap`  | double           | If on SURF or CRISP | "" if applicable |

- Where the default value is listed as `Algorithm` then the default value shall match that of the `ReflectometryReductionOneAuto` algorithm.

Fixes #21543. 



---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
